### ISSUE TITLE
Infinite loop with custom component object properties

### ISSUE THREAD

**Issue by andreixk** [vuejs/vue#4032]
### Vue.js version
2.0.3

### Reproduction Link
https://jsfiddle.net/vqq3shsp/
NOTE: the Loops number should be 1

### Steps to reproduce
* Create a custom component that is able to use v-bind. 
* Create a watcher that updates a local data variable based on the 'value' prop.
* Create a watcher on the local variable to emit the updated value back to the parent.
* Set the value of the local variable to an object.

### What is Expected?

The local value gets updated, an event is emitted once to the parent to update the bound 'value' prop.

### What is actually happening?

The watcher on 'value' keeps firing, which keeps calling the local variable watcher, which in turn updates the parent, which updates the 'value' once again and so on.

Basically, it should work like this: 
https://jsfiddle.net/zk5uzuLb/
-works fine if the v-model value is a string/int/bool, etc. The issue starts if the value is an object or an array.

I understand that in javascript ['a','b'] != ['a','b'], but it stands to reason that watcher shouldn't be called if new value and old value are equivalent (e.g. JSON.stringify(newValue) == JSON.stringify(oldValue)) or am I missing something here?

**Comment by HerringtonDarkholme** [comment]
Hi, IMHO this is a desired behavior as described [here](https://vuejs.org/guide/list.html#Array-Change-Detection), and you have noticed it.

> I understand that in javascript ['a','b'] != ['a','b']

Vue relies on object identity to tell whether change happens. This is a fundamental behavior that cannot be changed easily. It saves a lot of bytes of diffing algorithm and change detection strategy. 

Your pain point is understandable, but changing the behavior will incur more edge case (say, passing function as props), and more importantly, will introduce breaking change which is hard, if not impossible, for user to migrate.

**Comment by simplesmiler** [comment]
You can easily achieve this by adding a custom equality check to the start of the `someLocal` watcher:
`if (JSON.stringify(newValue) === JSON.stringify(oldValue)) return;`

**Comment by andreixk** [comment]
This works only for first update. However, if there is a subsequent update inside the component, it doesn't update the parent: https://jsfiddle.net/vqq3shsp/4/
